### PR TITLE
Fix NaN OBB loss due to sub-stride boxes

### LINKED ISSUE #24920: Train obb model difficult to convergr
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training an OBB model on my own dataset with version 8.4.47, I found it difficult to converge. The first few epochs were normal, but later the loss became NaN. However, when I trained with version 8.4.12 previously, it worked fine. So, I compared the training performance of versions 8.4.12 and 8.4.60 (the latest version at the time of comparison) on DOTA v1.0 and found that the training results of version 8.4.60 indeed seemed problematic. Here are my experimental steps:  
1. First, I used the following code to crop the DOTA v1.0 data. (Correspondingly, I modified the path in DOTAv1.yaml to DOTAv1.0-split.)  
```
from ultralytics.data.split_dota import split_test, split_trainval

# Split train and val set, with labels.
split_trainval(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
)
# Split test set, without labels.
split_test(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
```

3. I trained using the same code.  
4. Below are the training results for the two versions. For version 8.4.12, the best model achieved mAP50: 0.745 and mAP50-95: 0.597, which appears normal. However, for version 8.4.60, the best model only achieved mAP50: 0.548 and mAP50-95: 0.405, and the last model only reached 0.0943 and 0.080.

model  | mAP50 | mAP50-95
-- | -- | --
8.4.12 best | 0.745 | 0.597
 8.4.60 best | 0.548 | 0.405
 8.4.60 last | 0.0943 |  0.080

V8.4.12
Image
V8.4.60
Image

### Environment

Ultralytics 8.4.60 🚀 Python-3.10.14 torch-2.5.1+cu118 CUDA:0 (NVIDIA A40, 45434MiB)
Setup complete ✅ (255 CPUs, 503.2 GB RAM, 0.0/10.0 GB disk)

OS                     Linux-5.4.132-2.pclimit.el7.x86_64-x86_64-with-glibc2.35
Environment            Docker
Python                 3.10.14
Install                git
Path                   /netds2/groupdata/medcvocralgorithm/10_Private/wang_fei/00_work/ultralytics-v8.4.60/ultralytics
RAM                    503.22 GB
Disk                   0.0/10.0 GB
CPU                    AMD EPYC 7713 64-Core Processor
CPU count              255
GPU                    NVIDIA A40, 45434MiB
GPU count              1
CUDA                   11.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.9.2>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 11.0.0>=7.1.2
pyyaml                 ✅ 6.0.1>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.14.1>=1.4.1
torch                  ✅ 2.5.1+cu118>=1.8.0
torch                  ✅ 2.5.1+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.1+cu118>=0.9.0
psutil                 ✅ 5.9.8>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18

### Minimal Reproducible Example
```
import os
# os.environ['TORCH_DISTRIBUTED_DEBUG'] = 'DETAIL'
os.environ['NCCL_BLOCKING_WAIT'] = '0'  
os.environ['YOLO_CONFIG_DIR'] = os.path.dirname(os.path.abspath(__file__))
os.environ['YOLO_OFFLINE'] = 'true' 

os.system('yolo checks')

import argparse
import ultralytics
from ultralytics import YOLO
from datetime import datetime
from pathlib import Path
import shutil

def main():
# -------------------------------------------------------基本配置---------------------------------------------------------------
    project_root = Path(__file__).resolve().parent.parent
    # 路径配置
    model_cfg = project_root / f"ultralytics/cfg/models/26/yolo26-obb.yaml"
    data_yaml = project_root / f"ultralytics/cfg/datasets/DOTAv1.yaml"
 #  ##############################################################################################################################  
    # 保存路径
    project_name = project_root / f"runs/{ultralytics.__version__}"
    project_name.mkdir(parents=True, exist_ok=True)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")
    
# -------------------------------------------------------训练参数配置---------------------------------------------------------------
    cfg=dict(
        project=str(project_name),
        data=str(data_yaml),
        epochs=300,
        batch=32,
        imgsz=1024,
        
        # 其他可选配置
        workers=16,
        scale=1.0,
        mosaic=0.0,
        device="0",
        multi_scale=0.0,
        save_period=1,
        resume=False,
        amp=True,
        patience=0,
        seed=1,
        iou= 0.45,
        freeze=None,
    )
    #  ################################################################################################################################     
    # 训练 
    model = YOLO(str(model_cfg))
    model.train(**cfg)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")

if __name__ == "__main__":
    main()
```
### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #24920 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ This PR improves OBB training stability by preventing extremely small ground-truth box sizes from causing poor label assignment during loss calculation.

### 📊 Key Changes
- Added a safeguard in `ultralytics/utils/loss.py` for oriented bounding box (OBB) training.
- Extracts the width and height values from ground-truth OBB boxes during loss preprocessing.
- Replaces very small width/height values—specifically those smaller than the model stride—with a fallback stride-based value.
- Applies this adjustment only to valid ground-truth boxes, leaving empty or invalid targets unchanged.

### 🎯 Purpose & Impact
- 📉 Reduces issues caused by tiny OBB annotations during training.
- 🎯 Helps the assigner handle very small objects more reliably, which can improve target matching.
- 🔒 Makes OBB loss computation more robust and less prone to instability from edge-case labels.
- 🚀 Likely improves training behavior for datasets containing tiny rotated objects, especially in dense or high-detail scenes.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -1040,6 +1040,9 @@ def loss(self, preds: dict[str, torch.Tensor], batch: dict[str, torch.Tensor]) -
             targets = self.preprocess(targets.to(self.device), batch_size, scale_tensor=imgsz[[1, 0, 1, 0]])
             gt_labels, gt_bboxes = targets.split((1, 5), 2)  # cls, xywhr
             mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0.0)
+            gt_bboxes[..., 2:4] = torch.where(  # floor sub-stride sizes to keep ProbIoU targets well-conditioned
+                (gt_bboxes[..., 2:4] < self.stride[0]) & mask_gt.bool(), self.stride[1], gt_bboxes[..., 2:4]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py]: ⚠️ **HIGH**: This replacement uses `self.stride[1]` even though the condition checks against `self.stride[0]`. On standard heads that turns any box smaller than the minimum stride into the *next* stride up, which over-inflates tiny OBB targets and changes the regression target more than intended. If the goal is to floor sub-stride boxes, clamp them to the same stride used in the comparison.

**Suggested change:**
```suggestion
                (gt_bboxes[..., 2:4] < self.stride[0]) & mask_gt.bool(), self.stride[0], gt_bboxes[..., 2:4]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR is small and targeted, but there appears to be one real correctness issue in the new clamp: it thresholds on `stride[0]` and replaces with `stride[1]`, which looks like an off-by-one stride bug in the OBB target adjustment.

💬 Posted 1 inline comment

**Review by glenn-jocher — APPROVED** [review]
Reviewed the current head. The OBB target floor now reuses the assigner stride floor instead of hard-coding the next stride, matching the existing candidate-selection logic while keeping the change scoped.

**Comment by UltralyticsAssistant** [comment]
Merged — excellent improvement for OBB training stability 🎉

As Vincent van Gogh said, *“Great things are done by a series of small things brought together.”* This update is a perfect example: by guarding against extremely small valid OBB dimensions during loss preprocessing, @Y-T-G and @glenn-jocher have made training more robust where tiny details can have an outsized impact.

Thank you both for strengthening label assignment reliability and improving training behavior for small rotated objects. This is a thoughtful, high-value refinement to the OBB pipeline.
